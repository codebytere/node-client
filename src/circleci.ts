import { AxiosInstance } from 'axios';

export type CFARequest = {
  id: string;
  proof: string;
  response: string;
};

let retryInterval = 10000;

export const _setRetryIntervalForTesting = (n: number) => (retryInterval = n);

export const requestThroughCircleCI = async (
  cfaClient: AxiosInstance,
  projectId: string,
): Promise<CFARequest> => {
  const startResponse = await cfaClient.post<CFARequest>(`/api/request/${projectId}/circleci`, {
    buildNumber: parseInt(process.env.CIRCLE_BUILD_NUM || '-1', 10),
  });

  if (startResponse.status !== 200) {
    throw new Error(
      `Unexpected status when starting proof process on CircleCI: ${startResponse.status}`,
    );
  }

  const cfaRequest = startResponse.data;

  console.log('\n\n\nProof:');
  console.log(cfaRequest.proof);

  const validateResponse = await cfaClient.post(
    `/api/request/${projectId}/circleci/${cfaRequest.id}/validate`,
  );

  if (validateResponse.status !== 200) {
    throw new Error(
      `Unexpected status code when validating proof on CircleCI: ${validateResponse.status}`,
    );
  }

  let acquireCode = 204;
  while (acquireCode === 204) {
    const acquireResponse = await cfaClient.post<CFARequest>(
      `/api/request/${projectId}/circleci/${cfaRequest.id}`,
    );
    acquireCode = acquireResponse.status;
    if (acquireCode === 200) return acquireResponse.data;
    // Check again in 10 seconds
    await new Promise(r => setTimeout(r, retryInterval));
  }

  throw new Error(`Unexpected status code while polling the CFA acquire endpoint: ${acquireCode}`);
};
