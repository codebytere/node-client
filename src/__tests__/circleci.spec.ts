import { requestThroughCircleCI, _setRetryIntervalForTesting } from '../circleci';
import { AxiosInstance } from 'axios';

describe('getProofThroughCircleCI', () => {
  let mockClient: {
    post: jest.Mock;
  };
  const client = () => mockClient as any as AxiosInstance;
  let consoleSpy: jest.SpyInstance;

  beforeEach(() => {
    mockClient = {
      post: jest.fn(),
    };
    consoleSpy = jest.spyOn(console, 'log').mockImplementation(() => null);
    _setRetryIntervalForTesting(10);
  });

  afterEach(() => {
    _setRetryIntervalForTesting(10000);
  });

  it('should error out if the start request fails', async () => {
    mockClient.post.mockReturnValue(
      Promise.resolve({
        status: 401,
      }),
    );
    await expect(requestThroughCircleCI(client(), '1')).rejects.toMatchSnapshot();
  });

  it('should log out the provided proof content', async () => {
    mockClient.post.mockReturnValue(
      Promise.resolve({
        status: 200,
        data: {
          proof: 'abcdefg',
        },
      }),
    );
    await requestThroughCircleCI(client(), '1');
    expect(consoleSpy).toHaveBeenCalledTimes(2);
    expect(consoleSpy).toHaveBeenNthCalledWith(1, '\n\n\nProof:');
    expect(consoleSpy).toHaveBeenNthCalledWith(2, 'abcdefg');
  });

  it('should error out if the validation request fails', async () => {
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
        data: {
          proof: 'my_proof',
        },
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 403,
      }),
    );
    await expect(requestThroughCircleCI(client(), '1')).rejects.toMatchSnapshot();
  });

  it('should error out if the acquire request fails', async () => {
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
        data: {
          proof: 'my_proof',
        },
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 400,
      }),
    );
    await expect(requestThroughCircleCI(client(), '1')).rejects.toMatchSnapshot();
  });

  it('should eventually error out if one of the acquire requests fails', async () => {
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
        data: {
          proof: 'my_proof',
        },
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 204,
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 400,
      }),
    );
    await expect(requestThroughCircleCI(client(), '1')).rejects.toMatchSnapshot();
  });

  it('should keep trying to acquire until content is available', async () => {
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
        data: {
          proof: 'my_proof',
        },
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 204,
      }),
    );
    mockClient.post.mockReturnValueOnce(
      Promise.resolve({
        status: 200,
        data: {
          token: '12345',
        },
      }),
    );
    await expect(requestThroughCircleCI(client(), '1')).resolves.toEqual({ token: '12345' });
  });
});
