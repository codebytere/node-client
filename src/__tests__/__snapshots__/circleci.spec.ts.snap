// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProofThroughCircleCI should error out if the acquire request fails 1`] = `[Error: Unexpected status code while polling the CFA acquire endpoint: 400]`;

exports[`getProofThroughCircleCI should error out if the start request fails 1`] = `[Error: Unexpected status when starting proof process on CircleCI: 401]`;

exports[`getProofThroughCircleCI should error out if the validation request fails 1`] = `[Error: Unexpected status code when validating proof on CircleCI: 403]`;

exports[`getProofThroughCircleCI should eventually error out if one of the acquire requests fails 1`] = `[Error: Unexpected status code while polling the CFA acquire endpoint: 400]`;
